wpilib/wpilib/watchdog.py

```diff
@@ -75,18 +171,73 @@ def reset(self) -> None:
 
     def enable(self) -> None:
         """Enables the watchdog timer."""
-        self.startTime = Timer.getFPGATimestamp()
+        self._startTime = hal.getFPGATime()
         self._isExpired = False
-        self.epochs.clear()
-        self.notifier.startPeriodic(self.timeout)
+        self._epochs.clear()
+
+        with self._queueMutex:
+            self._isExpired = False
+
+            watchdogs = self._watchdogs
+            try:
+                watchdogs.remove(self)
+            except ValueError:
+                pass
+            else:
+                heapq.heapify(watchdogs)
+
+            self._expirationTime = self._startTime + self._timeout
+            heapq.heappush(watchdogs, self)
+            self._schedulerWaiter.notify_all()
 
     def disable(self) -> None:
-        """Disable the watchdog."""
-        self.notifier.stop()
-
-    def timeoutFunc(self) -> None:
-        if not self._isExpired:
-            logger.info("Watchdog not fed after %ss" % (self.timeout,))
-            self.callback()
-            self._isExpired = True
-            self.disable()
+        """Disables the watchdog timer."""
+        with self._queueMutex:
+            self._isExpired = False
+
+            watchdogs = self._watchdogs
+            try:
+                watchdogs.remove(self)
+            except ValueError:
+                pass
+            else:
+                heapq.heapify(watchdogs)
+            self._schedulerWaiter.notify_all()
+
+    @classmethod
+    def _schedulerFunc(cls) -> None:
+        getFPGATime = hal.getFPGATime
+
+        with cls._queueMutex:
+            while cls._keepAlive:
+                if cls._watchdogs:
+                    delta = cls._watchdogs[0]._expirationTime - getFPGATime()
+                    timedOut = cls._schedulerWaiter.wait(delta)
+
+                    if timedOut:
+                        if (
+                            not cls._watchdogs
+                            or cls._watchdogs[0]._expirationTime > getFPGATime()
+                        ):
+                            continue
+
+                        # If the condition variable timed out, that means a Watchdog
+                        # timeout has occurred, so call its timeout function.
+                        watchdog = cls._watchdogs[0]
+
+                        if not watchdog.suppressTimeoutMessage:
+                            logger.info(
+                                "Watchdog not fed after %.6fs", watchdog._timeout / 1e6
+                            )
+                        cls._queueMutex.release()
+                        watchdog._callback()
```

Comment:
Hm. I feel like we should catch exceptions + log them here instead of letting the world burn down if a callback throws.

Reply:
With java you would need to throw an unchecked exception otherwise you'll get a compile error. I feel this is one of the cases where we need to differ from them, as it's significantly easier to get an exception for Python.

Reply:
This is a scenario where you could potentially end up in an infinite loop of exceptions if we do decide to catch them though.

Reply:
Remove the offending watchdog if the exception is thrown?

Reply:
Hm, maybe. Code might get rather messy though.

How many teams would you expect to create their own Watchdog objects anyway?

Reply:
I'm more worried that some user setting causes our code to do something bad?, and they can't easily figure out what it is. Seems bad to kill all the watchdogs when it's just one bad apple.